v0/destinations/hs/transform.js

```diff
@@ -1,244 +1,609 @@
+  removeUndefinedAndNullValues,
+  getDestinationExternalID,
+  constructPayload,
+  removeUndefinedAndNullAndEmptyValues
 } = require("../../util");
 const {
-  ConfigCategory,
-  mappingConfig,
   BATCH_CONTACT_ENDPOINT,
-  MAX_BATCH_SIZE
+  MAX_BATCH_SIZE,
+  TRACK_ENDPOINT,
+  IDENTIFY_CREATE_UPDATE_CONTACT,
+  IDENTIFY_CREATE_NEW_CONTACT,
+  hsCommonConfigJson,
+  IDENTIFY_CRM_UPDATE_NEW_CONTACT,
+  IDENTIFY_CRM_CREATE_NEW_CONTACT,
+  MAX_BATCH_SIZE_CRM_CONTACT,
+  BATCH_IDENTIFY_CRM_CREATE_NEW_CONTACT,
+  BATCH_IDENTIFY_CRM_UPDATE_NEW_CONTACT,
+  mappingConfig,
+  ConfigCategory,
+  TRACK_CRM_ENDPOINT,
+  CRM_ASSOCIATION_V3
 } = require("./config");
-const { getEmailAndUpdatedProps } = require("./util");
-
-const hSIdentifyConfigJson = mappingConfig[ConfigCategory.IDENTIFY.name];
-
-let hubSpotPropertyMap = {};
-
-function getKey(key) {
-  let modifiedKey = key.toLowerCase();
-  modifiedKey = modifiedKey.replace(/\s/g, "_");
-  modifiedKey = modifiedKey.replace(/\./g, "_");
-  return modifiedKey;
-}
-
-function getTraits(message) {
-  let traits = getFieldValueFromMessage(message, "traits");
-  if (!traits || !Object.keys(traits).length) {
-    traits = message.properties;
+const {
+  getTraits,
+  getProperties,
+  getTransformedJSON,
+  getEmailAndUpdatedProps,
+  formatPropertyValueForIdentify,
+  searchContacts,
+  getCRMUpdatedProps,
+  getEventAndPropertiesFromConfig
+} = require("./util");
+
+// use legacy API
+/**
+ * Using New API
+ * Ref - https://developers.hubspot.com/docs/api/crm/contacts
+ * @param {*} message
+ * @param {*} destination
+ * @param {*} propertyMap
+ * @returns
+ */
+const processIdentify = async (message, destination, propertyMap) => {
+  const { Config } = destination;
+  const traits = getFieldValueFromMessage(message, "traits");
+  const mappedToDestination = get(message, MappedToDestinationKey);
+  // if mappedToDestination is set true, then add externalId to traits
+  if (mappedToDestination) {
+    addExternalIdToTraits(message);
   }
 
-  return traits;
-}
-
-async function getProperties(destination) {
-  if (!hubSpotPropertyMap.length) {
-    let response;
-    const { apiKey } = destination.Config;
-    const url = `https://api.hubapi.com/properties/v1/contacts/properties?hapikey=${apiKey}`;
-    try {
-      response = await axios.get(url);
-    } catch (err) {
-      // check if exists err.response && err.response.status else 500
-
-      if (err.response) {
-        throw new CustomError(
-          JSON.stringify(err.response.data) ||
-            JSON.stringify(err.response.statusText) ||
-            "Failed to get hubspot properties",
-          err.response.status || 500
-        );
-      }
-      throw new CustomError(
-        "Failed to get hubspot properties : indvalid response",
-        500
-      );
-    }
+  // if (!traits || !traits.email) {
+  //   throw new CustomError(
+  //     "[HS]:: Identify without email is not supported.",
+  //     400
+  //   );
+  // }
```

Comment:
We decided email will be mandatory for both new and old API specs, so let us keep a check for email with `getFieldValueFromMessage` and throw appropriate error otherwise

Reply:
as further discussed we are not placing this constraint